src/pmpo_utils.hpp

```diff
@@ -145,13 +153,249 @@ class Vec3d{
+      
+    KOKKOS_INLINE_FUNCTION
+    Vec4d operator*(double scalar) const {
+      return Vec4d(coords_[0] * scalar, coords_[1] * scalar, coords_[2] * scalar, coords_[3] * scalar);
+    }
+      
+    KOKKOS_INLINE_FUNCTION
+    double dot(const Vec4d& v) const {
+      return coords_[0] * v.coords_[0] + coords_[1] * v.coords_[1] + coords_[2] * v.coords_[2] + coords_[3] * v.coords_[3];
+    }
+      
+    KOKKOS_INLINE_FUNCTION
+    double magnitude() const {
+      return std::sqrt(coords_[0] * coords_[0] + coords_[1] * coords_[1] + coords_[2] * coords_[2] + coords_[3] * coords_[3]);
+    }
+};
+
+class Matrix4d {
+    
+  private:
+  
+    double** data_;
+    int rows_=4;
+    int cols_=4;
+   
+  public:
+			        
+    KOKKOS_INLINE_FUNCTION
+    Matrix4d(Vec4d v0, Vec4d v1, Vec4d v2, Vec4d v3){
+      data_ = new double*[rows_];
+  
+      for (int i=0; i<rows_; i++){
+        data_[i] = new double[cols_];
+      }
+
+      for (int i=0; i<cols_; i++){
+        data_[0][i] = v0[i];
+        data_[1][i] = v1[i];
+        data_[2][i] = v2[i];
+        data_[3][i] = v3[i];
+      }
+    }
+
+    //Destructor    
+    KOKKOS_INLINE_FUNCTION
+    ~Matrix4d(){
+      for (int i=0; i<4; i++){
+        delete[] data_[i];
+      }
+      delete[] data_;
+    }
+        
+    //Retrieval
+    KOKKOS_INLINE_FUNCTION
+    double& operator()(int i, int j) { return data_[i][j];}
+
+    //Scalar division operator overloading
+    KOKKOS_INLINE_FUNCTION
+    Matrix4d operator/(double scalar) const {
+      Matrix4d result(Vec4d(0, 0, 0, 0), Vec4d(0, 0, 0, 0), Vec4d(0, 0, 0, 0), Vec4d(0, 0, 0, 0));
+      for (int i = 0; i < rows_; i++) {
+        for (int j = 0; j < cols_; j++) {
+          result(i, j) = data_[i][j] / scalar;
+        }
+      }
+      return result;
+    }
+
+    // Function to calculate the Frobenius norm
+    KOKKOS_INLINE_FUNCTION
+    double frobeniusNorm() const {
+      double sum = 0.0; // Initialize sum of squares
+      for (int i = 0; i < rows_; i++) {
+        for (int j = 0; j < cols_; j++) {
+          sum += data_[i][j]* data_[i][j]; // Sum of squares
+        }
+      }
+      return std::sqrt(sum); // Return the square root of the sum
+    }
+
+    // Function to calculate the trace of the matrix
+    KOKKOS_INLINE_FUNCTION
+    double trace() const {
+      double traceSum = 0.0; // Initialize sum of diagonal elements
+      for (int i = 0; i < rows_; i++) {
+        traceSum += data_[i][i]; // Sum diagonal elements
+      }
+      return traceSum;
+    }
+    
+    //Function to regularize the matrix
+    KOKKOS_INLINE_FUNCTION
+    void regularize(double eps) {
+      for (int i = 1; i < rows_; i++) {
+        data_[i][i]=data_[i][i]+eps;
+      }
+    }
+				        
+};
+
+
 KOKKOS_INLINE_FUNCTION
 void initArray(Vec2d* arr, int n, Vec2d fill){
     for(int i=0; i<n; i++){
         arr[i] = fill;
     }
 }
 
+
+KOKKOS_INLINE_FUNCTION
+void CholeskySolve4d(Matrix4d& A, double* x){
```

Comment:
Is right hand side assumed here: [1,0,0,0]? If so, then we should reflect that in name? For example, `Cholesky4dSolve_UnitRHS` or `Cholesky4dSolve_EHat0RHS` ... these names are ok (not great and not bad)